components/engine_rocks/src/event_listener.rs

```diff
@@ -190,7 +190,7 @@ impl RocksPersistenceListener {
 impl rocksdb::EventListener for RocksPersistenceListener {
     fn on_memtable_sealed(&self, info: &MemTableInfo) {
         self.0
-            .on_memtable_sealed(info.cf_name().to_string(), info.earliest_seqno());
+            .on_memtable_sealed(info.cf_name().to_string(), info.first_seqno());
```

Comment:
Why merge can break the assumption of earliest_seqno that "any write with a larger(or equal) sequence number will be present in this memtable or a later memtable"?

Reply:
First off, the assumption is already broken in case you forgot. It shouldn't contain the "or equal" condition. Secondly, it's easy to make it work with `earliest_seqno`. After all, there's only less than 10 LOC involving updating seqno. But I didn't want to because first_seqno is easier to reason with. It has the benefits of (1) doesn't interact unexpectedly with the initial value (the initial value is set to 0, stands for uninit, while for earliest_seqno the initial value is already non-zero) (2) has a tighter bound, while for earliest_seqno (in its current status), it's harder to separate memtables with only one keys.

Reply:
Of cause I remember the assumption is broken (or ambiguous) already, I just don't understand why merge can trigger actual problem (that's why I start a new conversation instead of continue the last one). If it's a bug we should also set `earliest_seqno` to a correct value (at least not less than last flushed memtable largest_seqno).

Reply:
My main idea is that earliest_seqno is a broken concept. We have three choices (1) ignore the broken assumption (2) fix it to obey the assumption (3) relax the assumption.

(1) is what I have chosen. And since it's ignored, the new behavior introduced by merge is correct.

To choose (2), you will have to re-implement earliest_seqno to be identical as first_seqno (both its behavior and its code). I don't see the point of making a duplicate of something already there and working fine.

To choose (3), as I explained earlier, it's not as strict as first_seqno. It's hard to distinguish memtables with single key.

> I just don't understand why merge can trigger actual problem

For every newly opened DB, the earliest_seqno of the first memtable will be 0, regardless of the seqno of memtables flushed before opening.

For merge, this "special" memtable is placed at a position newer than all memtables from merge sources. So its earliest_seqno will be 0 too.

Reply:
@tabokie The discussion is quite long. Could you summarize these two questions:
1) What's the broken assumption? Is it "flushed largest_seqno equals to earliest_seqno of next memtable."? What's the correct one then.
2) So is to fair to say that earliest_seqno is not necessarily the smallest seqno of entries in the memtable. It can be smaller than the smallest seqno in that memtable.  However on the other side, the first_seqno is always the smallest seqno in the memtable.  I'm not sure if this assertion is correct.

Reply:
https://github.com/facebook/rocksdb/issues/11278 contains more details. The assumption (as commented in RocksDB code) is A's largest seqno < B's earliest_seqno <= B's smallest seqno.

> So is to fair to say that earliest_seqno is not necessarily the smallest seqno of entries in the memtable.

It is in fact smallest-1 in most common cases, in rare cases smaller than that.

> However on the other side, the first_seqno is always the smallest seqno in the memtable.

Its name and its documentation don't guarantee this (the first is ambiguous in concurrent insert scenario), but the implementation guarantees it.